src/pages/iv-items/forms/BarcodeForm.js

```diff
@@ -0,0 +1,171 @@
+
+  const { formik } = useForm({
+    enableReinitialize: true,
+    validateOnChange: true,
+    validationSchema: yup.object({
+      barcodes: yup
+        .array()
+        .of(
+          yup.object().shape({
+            barcode: yup.string().test(function (value) {
+              if (numRows > 1) {
+                return !!value
+              }
+
+              return true
+            })
+          })
+        )
+        .required()
+    }),
+    initialValues: {
+      barcodes: [
+        {
+          id: 1,
+          itemId: recordId,
+          barcode: '',
+          muId: '',
+          defaultQty: '',
+          muRef: ''
+        }
+      ]
+    },
+    onSubmit: values => {
+      const { barcodes } = values
+
+      const isRowEmpty = row => {
+        return !row.barcode
+      }
+
+      if (barcodes.length === 1 && isRowEmpty(barcodes[0])) {
+        formik.setValues({ barcodes: [] })
+        postdata([])
+      } else {
+        postdata(values)
+      }
+    }
+  })
+
+  const postdata = async obj => {
+    try {
+      const data = obj?.barcodes?.map(({ itemId, ...rest }) => ({
+        itemId: recordId,
+        ...rest
+      }))
+
+      const list = {
+        itemId: recordId,
+        barcodes: data || []
+      }
+
+      await postRequest({
+        extension: InventoryRepository.Barcode.set2,
+        record: JSON.stringify(list)
+      })
+
+      toast.success(platformLabels.Edited)
+      getData()
+    } catch (error) {}
+  }
+
+  const columns = [
+    {
+      component: 'textfield',
+      label: labels.barcode,
+      name: 'barcode'
+    },
+
+    {
+      component: 'resourcecombobox',
+      name: 'muId',
+      label: labels.msUnit,
+      props: {
+        endpointId: store._msId ? InventoryRepository.MeasurementUnit.qry : '',
+        parameters: `_msId=${store._msId}`,
+        valueField: 'recordId',
+        displayField: 'reference',
+        columnsInDropDown: [
+          { key: 'reference', value: 'Reference' },
+          { key: 'name', value: 'Name' }
+        ],
+        mapping: [
+          { from: 'recordId', to: 'muId' },
+          { from: 'reference', to: 'muRef' }
+        ]
+      }
+    },
+
+    {
+      component: 'numberfield',
+      name: 'defaultQty',
+      label: labels.defaultQty,
+      props: {
+        allowNegative: false
+      }
+    }
+  ]
+  async function getData() {
+    try {
+      const response = await getRequest({
+        extension: InventoryRepository.Barcode.qry,
+        parameters: `_itemId=${recordId}&_pageSize=50&_startAt=0`
+      })
+
+      const modifiedList = response.list?.map((category, index) => ({
+        ...category,
+        id: index + 1
+      }))
+
+      formik.setValues({ barcodes: modifiedList })
+    } catch (error) {}
+  }
+  useEffect(() => {
+    if (recordId) {
+      getData()
+    }
+  }, [recordId])
+
+  useEffect(() => {
+    setNumRows(formik.values.barcodes.length)
+  }, [formik.values.barcodes])
```

Comment:
can't we merge these two useEffects together?

Reply:
i need the setNumRows(formik.values.barcodes.length) every time the number of columns change so if i merge them and i change the columns number it will do the get data many time